lib/logstash/outputs/sqs.rb

```diff
@@ -1,136 +1,175 @@
-        )
-      end
-      buffer_initialize(
-        :max_items => @batch_events,
-        :max_interval => @batch_timeout,
-        :logger => @logger
-      )
-    end
+    @sqs = Aws::SQS::Client.new(aws_options_hash)
+
+    if @batch_events > 10
+      raise LogStash::ConfigurationError, 'The maximum batch size is 10 events'
+    elsif @batch_events < 1
+      raise LogStash::ConfigurationError, 'The batch size must be greater than 0'
+   end
 
     begin
-      @logger.debug("Connecting to AWS SQS queue '#{@queue}'...")
-      @sqs_queue = @sqs.queues.named(@queue)
-      @logger.info("Connected to AWS SQS queue '#{@queue}' successfully.")
-    rescue Exception => e
-      @logger.error("Unable to access SQS queue '#{@queue}': #{e.to_s}")
-    end # begin/rescue
-  end # def register
+      @logger.debug('Connecting to SQS queue', :queue => @queue, :region => region)
+      @queue_url = @sqs.get_queue_url(:queue_name => @queue)[:queue_url]
+      @logger.info('Connected to SQS queue successfully', :queue => @queue, :region => region)
+    rescue Aws::SQS::Errors::ServiceError => e
+      @logger.error('Failed to connect to SQS', :error => e)
+      raise LogStash::ConfigurationError, 'Verify the SQS queue name and your credentials'
+    end
+  end
 
   public
-  def receive(event)
-    if @batch
-      buffer_receive(event.to_json)
-      return
+  def multi_receive_encoded(encoded_events)
+    if @batch and @batch_events > 1
+      multi_receive_encoded_batch(encoded_events)
+    else
+      multi_receive_encoded_single(encoded_events)
     end
-    @sqs_queue.send_message(event.to_json)
-  end # def receive
+  end
 
-  # called from Stud::Buffer#buffer_flush when there are events to flush
-  def flush(events, close=false)
-    @sqs_queue.batch_send(events)
+  private
+  def multi_receive_encoded_batch(encoded_events)
+    bytes = 0
+    entries = []
+
+    # Split the events into multiple batches to ensure that no single batch
+    # exceeds `@message_max_size` bytes.
+    encoded_events.each_with_index do |encoded_event, index|
+      event, encoded = encoded_event
+
+      if encoded.bytesize > @message_max_size
+        @logger.warn('Message exceeds maximum length and will be dropped', :message_size => encoded.bytesize)
+        next
+      end
+
+      if entries.size >= @batch_events or (bytes + encoded.bytesize) > @message_max_size
+        send_message_batch(entries)
+
+        bytes = 0
+        entries = []
+      end
+
+      bytes += encoded.bytesize
+      entries.push(:id => index.to_s, :message_body => encoded)
+    end
+
+    send_message_batch(entries) unless entries.empty?
   end
 
-  public
-  def close
-    buffer_flush(:final => true)
-  end # def close
+  private
+  def multi_receive_encoded_single(encoded_events)
+    encoded_events.each do |encoded_event|
+      event, encoded = encoded_event
+
+      if encoded.bytesize > @message_max_size
```

Comment:
I don't like the code duplication between `multi_receive_encoded_single` and `multi_receive_encoded_batch`. 

You should be able to do this with just one function. It will be sending "batches" of 1 if `batch_len` is set to one, but it will be just one code path. I don't see any gains for making the decision early in `multi_receive_encoded`.